plugins/modules/dcnm_image_upgrade.py

```diff
@@ -0,0 +1,5313 @@
+                {
+                    "deviceName": "cvd-1312-leaf",
+                    "ipAddress": "172.22.150.103",
+                    "policyName": "KR5M",
+                    "module": 1,
+                    "name": null,
+                    "modelName": "N9K-C93180YC-EX",
+                    "moduleType": "IO FPGA",
+                    "oldVersion": "0x15",
+                    "newVersion": "0x15"
+                },
+                {
+                    "deviceName": "cvd-1312-leaf",
+                    "ipAddress": "172.22.150.103",
+                    "policyName": "KR5M",
+                    "module": 1,
+                    "name": null,
+                    "modelName": "N9K-C93180YC-EX",
+                    "moduleType": "MI FPGA",
+                    "oldVersion": "0x4",
+                    "newVersion": "0x04"
+                }
+            ],
+            "bException": false,
+            "exceptionReason": null
+        },
+        "installPacakges": null,
+        "errMessage": ""
+    }
+    """
+
+    def __init__(self, module):
+        super().__init__(module)
+        self.class_name = self.__class__.__name__
+        self._init_properties()
+
+    def _init_properties(self):
+        self.properties = {}
+        self.properties["epld"] = False
+        self.properties["issu"] = True
+        self.properties["ndfc_data"] = None
+        self.properties["ndfc_response"] = None
+        self.properties["ndfc_result"] = None
+        self.properties["package_install"] = False
+        self.properties["policy_name"] = None
+        self.properties["serial_number"] = None
+        self.properties["epld_modules"] = None
+
+    def refresh(self):
+        """
+        Refresh self.data with current install-options from NDFC
+        """
+        if self.policy_name is None:
+            msg = f"{self.class_name}.refresh: "
+            msg += "instance.policy_name must be set before "
+            msg += "calling refresh()"
+            self.module.fail_json(msg)
+        if self.serial_number is None:
+            msg = f"{self.class_name}.refresh: "
+            msg += f"instance.serial_number must be set before "
+            msg += f"calling refresh()"
+            self.module.fail_json(msg)
+
+        path = self.endpoints.install_options.get("path")
+        verb = self.endpoints.install_options.get("verb")
+        self._build_payload()
+        msg = f"REMOVE: {self.class_name}.refresh: "
+        msg += f"payload: {self.payload}"
+        self.log_msg(msg)
+        self.properties["ndfc_response"] = dcnm_send(
+            self.module, verb, path, data=json.dumps(self.payload)
+        )
+        msg = f"REMOVE: {self.class_name}.refresh: "
+        msg += f"ndfc_response: {self.ndfc_response}"
+        self.log_msg(msg)
+        self.properties["ndfc_result"] = self._handle_response(self.ndfc_response, verb)
+        # TODO:2 should error message contain full response or just DATA.error?
```

Comment:
At least have the relevant fields that aids in debugging

Reply:
This was addressed in a later commit.